### ISSUE TITLE
Bug: Negated ignore patterns not working with ignorePatterns

### ISSUE THREAD

**Issue by payapula** [eslint/eslint#15359]
### Environment

Node version: 12.16.1
npm version: 6.13.4
Local ESLint version: 7.11.0
Global ESLint version: NA
Operating System: mac

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
Node version: 12.16.1
npm version: 6.13.4
Local ESLint version: 7.11.0
Global ESLint version: NA
Operating System: mac
```
</details>

In the `.eslintrc` file, I tried to ignore specific files that ends with just `.test.tsx` and keep linting the files which ends with `*.tl.test.tsx`

For this I tried the glob pattern `**/!(*.tl).test.tsx`,  according to glob pattern specifications this should only select those files that ends with `*.test.tsx` but eslint is selecting all the files that ends with `*.test.tsx` including the ones `*.tl.test.tsx`.

For the proof, I tried this pattern in [Glob Tester](https://toools.cloud/miscellaneous/glob-tester) and this is the result expected:
![globtester](https://user-images.githubusercontent.com/7134153/143605637-d5e9ca9f-27dc-4b26-95a4-42a51ca7678c.png)

I tried the same pattern in this tool as well [globster](https://globster.xyz/)

![globster](https://user-images.githubusercontent.com/7134153/143605653-fa3260c3-7137-4850-984d-02af3de46098.png)

Also in this tool [Glob Tool](https://www.digitalocean.com/community/tools/glob) 
![globtool](https://user-images.githubusercontent.com/7134153/143605661-45848647-9f88-4920-9da7-f28ad019fa9c.png)

Everywhere I get the expected result but just in **eslint**, it is not working as expected.

```js
// File: .eslintrc
{
"ignorePatterns": ["**/!(*.tl).test.tsx"],
}
```

The above file is placed in the root directory, Please refer the screenshot from above tools to understand the folder structure.

```js
src/module/component.test.tsx
src/module/component.tl.test.tsx
.eslintrc
```

Other Patterns that I have already tried and failed are:

- `src/**/!(*.tl).{spec,test}.{js,jsx,ts,tsx}`
- `**/src/**/!(*.tl).{spec,test}.{js,jsx,ts,tsx}`
- `!(*.tl).test.tsx`

### What did you expect to happen?

Only the files ending with `test.tsx` should be ignored

### What actually happened?

All the files ending with `test.tsx` are ignored, including the negated pattern `tl` like `component.tl.test.tsx` 

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

Let me know if any additional information is required to analyze the issue. 

This is kind of similar to the issues reported below:

- #5979
- #7209
- [ESLint "negated ignore pattern" not working](https://github.com/okonet/lint-staged/issues/204)

**Comment by mdjermanovic** [comment]
Hi @payapula, thanks for the issue!

 `"ignorePatterns"` feature works per the [.gitignore specification](https://git-scm.com/docs/gitignore).

> Everywhere I get the expected result but just in **eslint**, it is not working as expected.
> 
> ```js
> // File: .eslintrc
> {
> "ignorePatterns": ["**/!(*.tl).test.tsx"],
> }
> ```

This is expected behavior for `.gitignore` patterns. The negated pattern has no effect, because it is re-including files that were not ignored.

> ignore specific files that ends with just `.test.tsx` and keep linting the files which ends with `*.tl.test.tsx`

Can you try with the following configuration:

> ```js
> // File: .eslintrc
> {
> "ignorePatterns": ["*.test.tsx", "!*.tl.test.tsx"],
> }
> ```

**Comment by payapula** [comment]
Hi @mdjermanovic  

Thanks for the very quick reply 😀

It's my bad, you explained it wonderfully. I understand now, and Yes, the solution you suggested worked! 

```
// File: .eslintrc
{
"ignorePatterns": ["*.test.tsx", "!*.tl.test.tsx"],
}
```

Thank you so much for helping out 🎉
